File: gluecodium/src/test/resources/smoke/defaults/output/swift/smoke/StructWithEnums.swift
//
//
import Foundation
public struct StructWithEnums {
    public static let firstConstant: SomethingEnum = SomethingEnum.reallyFirst
    public var firstField: SomethingEnum
    public var explicitField: SomethingEnum
    public var lastField: SomethingEnum
    public init(firstField: SomethingEnum = SomethingEnum.reallyFirst, explicitField: SomethingEnum = SomethingEnum.explicit, lastField: SomethingEnum = SomethingEnum.last) {
        self.firstField = firstField
        self.explicitField = explicitField
        self.lastField = lastField
    }
    internal init(cHandle: _baseRef) {
        firstField = moveFromCType(smoke_StructWithEnums_firstField_get(cHandle))
        explicitField = moveFromCType(smoke_StructWithEnums_explicitField_get(cHandle))
        lastField = moveFromCType(smoke_StructWithEnums_lastField_get(cHandle))
    }
}
internal func copyFromCType(_ handle: _baseRef) -> StructWithEnums {
    return StructWithEnums(cHandle: handle)
}
internal func moveFromCType(_ handle: _baseRef) -> StructWithEnums {
    defer {
        smoke_StructWithEnums_release_handle(handle)
    }
    return copyFromCType(handle)
}
internal func copyToCType(_ swiftType: StructWithEnums) -> RefHolder {
    let c_firstField = moveToCType(swiftType.firstField)
    let c_explicitField = moveToCType(swiftType.explicitField)
    let c_lastField = moveToCType(swiftType.lastField)
    return RefHolder(smoke_StructWithEnums_create_handle(c_firstField.ref, c_explicitField.ref, c_lastField.ref))
}
internal func moveToCType(_ swiftType: StructWithEnums) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_StructWithEnums_release_handle)
}
internal func copyFromCType(_ handle: _baseRef) -> StructWithEnums? {
    guard handle != 0 else {
        return nil
    }
    let unwrappedHandle = smoke_StructWithEnums_unwrap_optional_handle(handle)
    return StructWithEnums(cHandle: unwrappedHandle) as StructWithEnums
}
internal func moveFromCType(_ handle: _baseRef) -> StructWithEnums? {
    defer {
        smoke_StructWithEnums_release_optional_handle(handle)
    }
    return copyFromCType(handle)
}
internal func copyToCType(_ swiftType: StructWithEnums?) -> RefHolder {
    guard let swiftType = swiftType else {
        return RefHolder(0)
    }
    let c_firstField = moveToCType(swiftType.firstField)
    let c_explicitField = moveToCType(swiftType.explicitField)
    let c_lastField = moveToCType(swiftType.lastField)
    return RefHolder(smoke_StructWithEnums_create_optional_handle(c_firstField.ref, c_explicitField.ref, c_lastField.ref))
}
internal func moveToCType(_ swiftType: StructWithEnums?) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_StructWithEnums_release_optional_handle)
}

File: gluecodium/src/test/resources/smoke/defaults/output/dart/lib/src/smoke/struct_with_enums.dart
import 'dart:ffi';
import 'package:library/src/_library_context.dart' as __lib;
import 'package:library/src/smoke/something_enum.dart';
class StructWithEnums {
  SomethingEnum firstField;
  SomethingEnum explicitField;
  SomethingEnum lastField;
  StructWithEnums._(this.firstField, this.explicitField, this.lastField);
  StructWithEnums()
    : firstField = SomethingEnum.reallyFirst, explicitField = SomethingEnum.explicit, lastField = SomethingEnum.last;
  static final SomethingEnum firstConstant = SomethingEnum.reallyFirst;
}
// StructWithEnums "private" section, not exported.
final _smokeStructwithenumsCreateHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Uint32, Uint32, Uint32),
    Pointer<Void> Function(int, int, int)
  >('library_smoke_StructWithEnums_create_handle'));
final _smokeStructwithenumsReleaseHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('library_smoke_StructWithEnums_release_handle'));
final _smokeStructwithenumsGetFieldfirstField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Uint32 Function(Pointer<Void>),
    int Function(Pointer<Void>)
  >('library_smoke_StructWithEnums_get_field_firstField'));
final _smokeStructwithenumsGetFieldexplicitField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Uint32 Function(Pointer<Void>),
    int Function(Pointer<Void>)
  >('library_smoke_StructWithEnums_get_field_explicitField'));
final _smokeStructwithenumsGetFieldlastField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Uint32 Function(Pointer<Void>),
    int Function(Pointer<Void>)
  >('library_smoke_StructWithEnums_get_field_lastField'));
Pointer<Void> smokeStructwithenumsToFfi(StructWithEnums value) {
  final _firstFieldHandle = smokeSomethingenumToFfi(value.firstField);
  final _explicitFieldHandle = smokeSomethingenumToFfi(value.explicitField);
  final _lastFieldHandle = smokeSomethingenumToFfi(value.lastField);
  final _result = _smokeStructwithenumsCreateHandle(_firstFieldHandle, _explicitFieldHandle, _lastFieldHandle);
  smokeSomethingenumReleaseFfiHandle(_firstFieldHandle);
  smokeSomethingenumReleaseFfiHandle(_explicitFieldHandle);
  smokeSomethingenumReleaseFfiHandle(_lastFieldHandle);
  return _result;
}
StructWithEnums smokeStructwithenumsFromFfi(Pointer<Void> handle) {
  final _firstFieldHandle = _smokeStructwithenumsGetFieldfirstField(handle);
  final _explicitFieldHandle = _smokeStructwithenumsGetFieldexplicitField(handle);
  final _lastFieldHandle = _smokeStructwithenumsGetFieldlastField(handle);
  try {
    return StructWithEnums._(
      smokeSomethingenumFromFfi(_firstFieldHandle),
      smokeSomethingenumFromFfi(_explicitFieldHandle),
      smokeSomethingenumFromFfi(_lastFieldHandle)
    );
  } finally {
    smokeSomethingenumReleaseFfiHandle(_firstFieldHandle);
    smokeSomethingenumReleaseFfiHandle(_explicitFieldHandle);
    smokeSomethingenumReleaseFfiHandle(_lastFieldHandle);
  }
}
void smokeStructwithenumsReleaseFfiHandle(Pointer<Void> handle) => _smokeStructwithenumsReleaseHandle(handle);
// Nullable StructWithEnums
final _smokeStructwithenumsCreateHandleNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_StructWithEnums_create_handle_nullable'));
final _smokeStructwithenumsReleaseHandleNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('library_smoke_StructWithEnums_release_handle_nullable'));
final _smokeStructwithenumsGetValueNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_StructWithEnums_get_value_nullable'));
Pointer<Void> smokeStructwithenumsToFfiNullable(StructWithEnums? value) {
  if (value == null) return Pointer<Void>.fromAddress(0);
  final _handle = smokeStructwithenumsToFfi(value);
  final result = _smokeStructwithenumsCreateHandleNullable(_handle);
  smokeStructwithenumsReleaseFfiHandle(_handle);
  return result;
}
StructWithEnums? smokeStructwithenumsFromFfiNullable(Pointer<Void> handle) {
  if (handle.address == 0) return null;
  final _handle = _smokeStructwithenumsGetValueNullable(handle);
  final result = smokeStructwithenumsFromFfi(_handle);
  smokeStructwithenumsReleaseFfiHandle(_handle);
  return result;
}
void smokeStructwithenumsReleaseFfiHandleNullable(Pointer<Void> handle) =>
  _smokeStructwithenumsReleaseHandleNullable(handle);
// End of StructWithEnums "private" section.
